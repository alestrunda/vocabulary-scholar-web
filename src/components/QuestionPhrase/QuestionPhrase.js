import React from 'react';
import PropTypes from 'prop-types';
import TextField from 'material-ui/TextField';
import Hint from '../Hint';

class QuestionPhrase extends React.Component {
  constructor(props) {
    super(props);
    this.inputAnswerRef = React.createRef();
  }

  state = {
    field: '',
    error: '',
  };

  componentDidMount() {
    this.handleQuestionReady();
  }

  componentDidUpdate(prevProps) {
    //it's important to trigger questionReady, do it when word changes and when phrase is set - it could be set later, that's why comparing with prevProps
    if (
      this.props.wordID !== prevProps.wordID ||
      (!prevProps.phrase && this.props.phrase)
    ) {
      this.setState(
        {
          field: '',
        },
        () => this.handleQuestionReady()
      );
    }
  }

  handleQuestionReady() {
    if (!this.props.phrase) return;
    if (this.inputAnswerRef.current) this.inputAnswerRef.current.focus();
    this.props.onQuestionReady && this.props.onQuestionReady();
  }

  handleFieldChange = event => {
    this.setState({
      field: event.target.value,
    });
    this.props.onFieldChange && this.props.onFieldChange(event.target.value);
  };

  handleSubmit = e => {
    e.preventDefault();
    this.props.onSubmit && this.props.onSubmit(e);
  };

  renderPhrase() {
    return (
      <form className="phrase-word-question" onSubmit={this.handleSubmit}>
        {//render first part of the phrase
        this.props.phrase.substring(0, this.props.startIndex)}
        <TextField
          ref={this.inputAnswerRef}
          id={this.props.wordID}
          disabled={this.props.disabled}
          autoComplete="off"
          style={{
            maxWidth: 120,
            marginLeft: 10,
            marginRight: 10,
            verticalAlign: 'middle',
          }}
          inputStyle={{
            fontSize: 24,
            textAlign: 'center',
          }}
          onChange={this.handleFieldChange}
          value={this.state.field}
          errorText={this.props.error}
        />
        {
          //render hint
          <Hint
            className="mr5"
            hint={this.props.hint}
            hintType={this.props.hintType}
            wordID={this.props.wordID}
          />
        }
        {//render second part of the phrase
        this.props.phrase.substring(
          this.props.endIndex,
          this.props.phrase.length
        )}
      </form>
    );
  }

  render() {
    const isPhraseAvailable =
      this.props.startIndex !== undefined &&
      this.props.endIndex !== undefined &&
      this.props.endIndex !== 0 &&
      this.props.startIndex !== this.props.endIndex &&
      this.props.phrase;
    if (this.props.hideIfNotAvailable && !isPhraseAvailable) return null;
    return (
      <div className={this.props.className} style={this.props.style}>
        {isPhraseAvailable ? this.renderPhrase() : 'Not available'}
      </div>
    );
  }
}

QuestionPhrase.propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
  word: PropTypes.string.isRequired,
  wordID: PropTypes.string.isRequired,
  phrase: PropTypes.string,
  startIndex: PropTypes.number,
  endIndex: PropTypes.number,
  onQuestionReady: PropTypes.func,
  onFieldChange: PropTypes.func,
  onSubmit: PropTypes.func,
  disabled: PropTypes.bool,
  hideIfNotAvailable: PropTypes.bool,
  error: PropTypes.string,
  hint: PropTypes.string,
  hintType: PropTypes.string,
};

QuestionPhrase.defaultProps = {
  disabled: false,
  hideIfNotAvailable: false,
  error: '',
};

export default QuestionPhrase;
