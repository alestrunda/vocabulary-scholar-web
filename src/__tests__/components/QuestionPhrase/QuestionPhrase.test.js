import React from 'react';
import { shallow } from 'enzyme';

import QuestionPhrase from '../../../components/QuestionPhrase';

it('renders not available', () => {
  const word = 'abc',
    wordID = 'abc_123',
    phrase = 'Lorem ipsum dolor';
  const root = shallow(
    <QuestionPhrase word={word} wordID={wordID} phrase={phrase} />
  );
  expect(root).toMatchSnapshot();
});

it('renders correctly', () => {
  const word = 'abc',
    wordID = 'abc_123',
    phrase = 'Lorem ipsum dolor',
    startIndex = 6,
    endIndex = 11;
  const root = shallow(
    <QuestionPhrase
      word={word}
      wordID={wordID}
      phrase={phrase}
      startIndex={startIndex}
      endIndex={endIndex}
    />
  );
  expect(root).toMatchSnapshot();
});

it('renders correctly when disabled', () => {
  const word = 'abc',
    wordID = 'abc_123',
    phrase = 'Lorem ipsum dolor',
    startIndex = 6,
    endIndex = 11;
  const root = shallow(
    <QuestionPhrase
      word={word}
      wordID={wordID}
      phrase={phrase}
      startIndex={startIndex}
      endIndex={endIndex}
      disabled
    />
  );
  expect(root).toMatchSnapshot();
});
