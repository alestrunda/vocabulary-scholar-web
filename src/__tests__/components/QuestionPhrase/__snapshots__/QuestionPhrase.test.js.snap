// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionPhrase
    disabled={false}
    endIndex={11}
    error=""
    hideIfNotAvailable={false}
    phrase="Lorem ipsum dolor"
    startIndex={6}
    word="abc"
    wordID="abc_123"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        className="phrase-word-question"
        onSubmit={[Function]}
      >
        Lorem 
        <TextField
          autoComplete="off"
          disabled={false}
          errorText=""
          floatingLabelFixed={false}
          fullWidth={false}
          id="abc_123"
          inputStyle={
            Object {
              "fontSize": 24,
              "textAlign": "center",
            }
          }
          multiLine={false}
          onChange={[Function]}
          rows={1}
          style={
            Object {
              "marginLeft": 10,
              "marginRight": 10,
              "maxWidth": 120,
              "verticalAlign": "middle",
            }
          }
          type="text"
          underlineShow={true}
          value=""
        />
        <Hint
          className="mr5"
          wordID="abc_123"
        />
         dolor
      </form>,
      "className": undefined,
      "style": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Lorem ",
          <TextField
            autoComplete="off"
            disabled={false}
            errorText=""
            floatingLabelFixed={false}
            fullWidth={false}
            id="abc_123"
            inputStyle={
              Object {
                "fontSize": 24,
                "textAlign": "center",
              }
            }
            multiLine={false}
            onChange={[Function]}
            rows={1}
            style={
              Object {
                "marginLeft": 10,
                "marginRight": 10,
                "maxWidth": 120,
                "verticalAlign": "middle",
              }
            }
            type="text"
            underlineShow={true}
            value=""
          />,
          <Hint
            className="mr5"
            wordID="abc_123"
          />,
          " dolor",
        ],
        "className": "phrase-word-question",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        "Lorem ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": "off",
            "disabled": false,
            "errorText": "",
            "floatingLabelFixed": false,
            "fullWidth": false,
            "id": "abc_123",
            "inputStyle": Object {
              "fontSize": 24,
              "textAlign": "center",
            },
            "multiLine": false,
            "onChange": [Function],
            "rows": 1,
            "style": Object {
              "marginLeft": 10,
              "marginRight": 10,
              "maxWidth": 120,
              "verticalAlign": "middle",
            },
            "type": "text",
            "underlineShow": true,
            "value": "",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "mr5",
            "hint": undefined,
            "hintType": undefined,
            "wordID": "abc_123",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        " dolor",
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          className="phrase-word-question"
          onSubmit={[Function]}
        >
          Lorem 
          <TextField
            autoComplete="off"
            disabled={false}
            errorText=""
            floatingLabelFixed={false}
            fullWidth={false}
            id="abc_123"
            inputStyle={
              Object {
                "fontSize": 24,
                "textAlign": "center",
              }
            }
            multiLine={false}
            onChange={[Function]}
            rows={1}
            style={
              Object {
                "marginLeft": 10,
                "marginRight": 10,
                "maxWidth": 120,
                "verticalAlign": "middle",
              }
            }
            type="text"
            underlineShow={true}
            value=""
          />
          <Hint
            className="mr5"
            wordID="abc_123"
          />
           dolor
        </form>,
        "className": undefined,
        "style": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Lorem ",
            <TextField
              autoComplete="off"
              disabled={false}
              errorText=""
              floatingLabelFixed={false}
              fullWidth={false}
              id="abc_123"
              inputStyle={
                Object {
                  "fontSize": 24,
                  "textAlign": "center",
                }
              }
              multiLine={false}
              onChange={[Function]}
              rows={1}
              style={
                Object {
                  "marginLeft": 10,
                  "marginRight": 10,
                  "maxWidth": 120,
                  "verticalAlign": "middle",
                }
              }
              type="text"
              underlineShow={true}
              value=""
            />,
            <Hint
              className="mr5"
              wordID="abc_123"
            />,
            " dolor",
          ],
          "className": "phrase-word-question",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          "Lorem ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoComplete": "off",
              "disabled": false,
              "errorText": "",
              "floatingLabelFixed": false,
              "fullWidth": false,
              "id": "abc_123",
              "inputStyle": Object {
                "fontSize": 24,
                "textAlign": "center",
              },
              "multiLine": false,
              "onChange": [Function],
              "rows": 1,
              "style": Object {
                "marginLeft": 10,
                "marginRight": 10,
                "maxWidth": 120,
                "verticalAlign": "middle",
              },
              "type": "text",
              "underlineShow": true,
              "value": "",
            },
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "mr5",
              "hint": undefined,
              "hintType": undefined,
              "wordID": "abc_123",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          " dolor",
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders correctly when disabled 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionPhrase
    disabled={true}
    endIndex={11}
    error=""
    hideIfNotAvailable={false}
    phrase="Lorem ipsum dolor"
    startIndex={6}
    word="abc"
    wordID="abc_123"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        className="phrase-word-question"
        onSubmit={[Function]}
      >
        Lorem 
        <TextField
          autoComplete="off"
          disabled={true}
          errorText=""
          floatingLabelFixed={false}
          fullWidth={false}
          id="abc_123"
          inputStyle={
            Object {
              "fontSize": 24,
              "textAlign": "center",
            }
          }
          multiLine={false}
          onChange={[Function]}
          rows={1}
          style={
            Object {
              "marginLeft": 10,
              "marginRight": 10,
              "maxWidth": 120,
              "verticalAlign": "middle",
            }
          }
          type="text"
          underlineShow={true}
          value=""
        />
        <Hint
          className="mr5"
          wordID="abc_123"
        />
         dolor
      </form>,
      "className": undefined,
      "style": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Lorem ",
          <TextField
            autoComplete="off"
            disabled={true}
            errorText=""
            floatingLabelFixed={false}
            fullWidth={false}
            id="abc_123"
            inputStyle={
              Object {
                "fontSize": 24,
                "textAlign": "center",
              }
            }
            multiLine={false}
            onChange={[Function]}
            rows={1}
            style={
              Object {
                "marginLeft": 10,
                "marginRight": 10,
                "maxWidth": 120,
                "verticalAlign": "middle",
              }
            }
            type="text"
            underlineShow={true}
            value=""
          />,
          <Hint
            className="mr5"
            wordID="abc_123"
          />,
          " dolor",
        ],
        "className": "phrase-word-question",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        "Lorem ",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": "off",
            "disabled": true,
            "errorText": "",
            "floatingLabelFixed": false,
            "fullWidth": false,
            "id": "abc_123",
            "inputStyle": Object {
              "fontSize": 24,
              "textAlign": "center",
            },
            "multiLine": false,
            "onChange": [Function],
            "rows": 1,
            "style": Object {
              "marginLeft": 10,
              "marginRight": 10,
              "maxWidth": 120,
              "verticalAlign": "middle",
            },
            "type": "text",
            "underlineShow": true,
            "value": "",
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "className": "mr5",
            "hint": undefined,
            "hintType": undefined,
            "wordID": "abc_123",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        " dolor",
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          className="phrase-word-question"
          onSubmit={[Function]}
        >
          Lorem 
          <TextField
            autoComplete="off"
            disabled={true}
            errorText=""
            floatingLabelFixed={false}
            fullWidth={false}
            id="abc_123"
            inputStyle={
              Object {
                "fontSize": 24,
                "textAlign": "center",
              }
            }
            multiLine={false}
            onChange={[Function]}
            rows={1}
            style={
              Object {
                "marginLeft": 10,
                "marginRight": 10,
                "maxWidth": 120,
                "verticalAlign": "middle",
              }
            }
            type="text"
            underlineShow={true}
            value=""
          />
          <Hint
            className="mr5"
            wordID="abc_123"
          />
           dolor
        </form>,
        "className": undefined,
        "style": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Lorem ",
            <TextField
              autoComplete="off"
              disabled={true}
              errorText=""
              floatingLabelFixed={false}
              fullWidth={false}
              id="abc_123"
              inputStyle={
                Object {
                  "fontSize": 24,
                  "textAlign": "center",
                }
              }
              multiLine={false}
              onChange={[Function]}
              rows={1}
              style={
                Object {
                  "marginLeft": 10,
                  "marginRight": 10,
                  "maxWidth": 120,
                  "verticalAlign": "middle",
                }
              }
              type="text"
              underlineShow={true}
              value=""
            />,
            <Hint
              className="mr5"
              wordID="abc_123"
            />,
            " dolor",
          ],
          "className": "phrase-word-question",
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          "Lorem ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoComplete": "off",
              "disabled": true,
              "errorText": "",
              "floatingLabelFixed": false,
              "fullWidth": false,
              "id": "abc_123",
              "inputStyle": Object {
                "fontSize": 24,
                "textAlign": "center",
              },
              "multiLine": false,
              "onChange": [Function],
              "rows": 1,
              "style": Object {
                "marginLeft": 10,
                "marginRight": 10,
                "maxWidth": 120,
                "verticalAlign": "middle",
              },
              "type": "text",
              "underlineShow": true,
              "value": "",
            },
            "ref": Object {
              "current": null,
            },
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "mr5",
              "hint": undefined,
              "hintType": undefined,
              "wordID": "abc_123",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          " dolor",
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;

exports[`renders not available 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <QuestionPhrase
    disabled={false}
    error=""
    hideIfNotAvailable={false}
    phrase="Lorem ipsum dolor"
    word="abc"
    wordID="abc_123"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "Not available",
      "className": undefined,
      "style": undefined,
    },
    "ref": null,
    "rendered": "Not available",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "Not available",
        "className": undefined,
        "style": undefined,
      },
      "ref": null,
      "rendered": "Not available",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
